Add a widget for editing file names

evsh opened the pull request:
The combinations of `QLineEdit` or `QComboBox` with a browse button for file system paths editing most likely results in selecting files or dirs via the button and a file dialog. However,  typing with auto completion might provide a better alternative. This PR implements it in the standard for Qt way, using `QCompleter` and `QFileSystemModel`.

The PR adds `FileSystemPathLineEdit` and `FileSystemPathComboEdit` classes (share most of the implementation) that provides widgets for editing file names with `QLineEdit` or `QComboBox` (in editable mode). The widgets also contain the "Browse" button, which launches a file dialog. Properties of the dialog (title, file filter, etc.) are specified as `FileSystemPathEdit` class properties. 

The class also uses `Utils::Fs::fromNativePath()` and `Utils::Fs::toNativePath()` internally thus eliminating the need to use them with each widget explicitly. 

Screenshots show example for Add torrent dialog:

![qbt-adddialog-completion](https://cloud.githubusercontent.com/assets/394621/16087605/dc0b9e68-3323-11e6-862c-3e3b3cbd15b1.png)
![qbt-adddialog-completion-menu](https://cloud.githubusercontent.com/assets/394621/16087629/f516180c-3323-11e6-8cf9-8823862e35eb.png)

glassez commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,365 @@
+#endif
+}
+
+#include "moc_fspathedit.cpp"
```
Why do you include this file?

evsh replied:
Due to  `Q_PRIVATE_SLOT(d_func(), void browseActionTriggered())` in the `FileSystemPathEdit` class.

evsh replied:
Q_PRIVATE_SLOT replaced with lambda.

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -1391,73 +1401,10 @@ QString OptionsDialog::askForExportDir(const QString& currentExportPath)
-{
-    QDir lastDir(Utils::Fs::fromNativePath(m_ui->textFilterPath->text()));
-    QString lastPath = lastDir.exists() ? lastDir.absolutePath() : QDir::homePath();
-    QString newPath = QFileDialog::getOpenFileName(this, tr("Choose an IP filter file"), lastPath, tr("All supported filters") + QString(" (*.dat *.p2p *.p2b);;.dat (*.dat);;.p2p (*.p2p);;.p2b (*.p2b)"));
```
Is it possible to preserve?
```c++
QString(" (*.dat *.p2p *.p2b);;.dat (*.dat);;.p2p (*.p2p);;.p2b (*.p2b)");
```

evsh replied:
Thanks! How could I lose half of the filter?..

evsh replied:
Done.

Chocobo1 commented:
Is it a good idea to move the new code to its own folder `gui/fspathedit`? There are already 99 files + dirs in `gui`...

glassez commented:
>Is it a good idea to move the new code to its own folder  `gui/fspathedit`? 

IMO, it's good.

evsh commented:
> Is it a good idea to move the new code to its own folder `gui/fspathedit`?

Maybe `gui/widgets` is a better name?

glassez commented:
>Maybe `gui/widgets` is a better name?

Then we should move all other widgets there.

evsh commented:
... and do the same for  `gui/views`, `gui/models`, and, `gui/dialogs`.

LordNyriox commented:
@sledgehammer999 approveFeature

glassez commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -382,23 +385,20 @@ void AddNewTorrentDialog::updateDiskSpaceLabel()
 }
 
-void AddNewTorrentDialog::onSavePathChanged(int index)
+void AddNewTorrentDialog::onSavePathChanged(const QString &/*text*/)
```
You can connect slots with fewer parameters than signal has.

evsh replied:
Excuse me, I don't understand what did you mean here.

glassez replied:
I mean you can turn it to `void AddNewTorrentDialog::onSavePathChanged()` unless you need `QString` param.

evsh replied:
Ah, thanks.

glassez commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -658,6 +623,15 @@ void AddNewTorrentDialog::accept()
 
     m_torrentGuard->markAsAddedToSession();
+
+    // add selected save path to the save path history
```
I'm confused about it outside of class. Why not do this in the destructor?

glassez commented on src/gui/fspathedit.h:
```diff
@@ -0,0 +1,151 @@
+
+    enum class Mode
+    {
+        FileOpen,                                 //!< Working mode for opening files, shows open dialog
```
Would you like to put these comments closer to the enum items?

glassez commented on src/gui/fspathedit.h:
```diff
@@ -0,0 +1,151 @@
+protected slots:
+    void onPathEdited();
+
+signals:
```
Public members should appear before protected/private. Signals are public.
Also protected/private slots is less protected/private than other protected/private members.

glassez commented on src/gui/fspathedit.h:
```diff
@@ -0,0 +1,151 @@
+    Q_DISABLE_COPY(FileSystemPathEdit)
+    class FileSystemPathEditPrivate;
+    Q_DECLARE_PRIVATE(FileSystemPathEdit)
+    Q_PRIVATE_SLOT(d_func(), void browseActionTriggered())
```
Qt has a naming convention for private slots (don't remember exactly how it looks).

evsh replied:
Replaced with lambda.

glassez commented on src/gui/fspathedit.h:
```diff
@@ -0,0 +1,151 @@
+    class FileSystemPathEditPrivate;
+    Q_DECLARE_PRIVATE(FileSystemPathEdit)
+    Q_PRIVATE_SLOT(d_func(), void browseActionTriggered())
+    QScopedPointer<FileSystemPathEditPrivate> const d_ptr;
```
Why do you use QScopedPointers in these places? Why not usual Qt ownership?

evsh replied:
The private class does not inherit QObject.

glassez commented on src/gui/fspathedit_p.cpp:
```diff
@@ -0,0 +1,136 @@
+    , m_completer {new QCompleter(this)}
+    , m_browseAction {nullptr}
+{
+#if !defined QBT_USES_QT5 && defined Q_OS_WIN
```
QBT_USES_QT5 is outdated.

glassez commented on src/gui/fspathedit_p.h:
```diff
@@ -0,0 +1,101 @@
+    {
+    public:
+        virtual ~FileEditorWithCompletion() = default;
+        virtual void completeDirectoriesOnly(bool completeDirsOnly) = 0;
```
I wonder why C++ still does not have `abstract` keyword for this purpose?

evsh commented:
When I update this PR now, do you think I should move its files into `gui/widgets` or not? Should I submit another PR instead to rearrange all files (`gui/models`, `gui/views`, `gui/widgets`) and break basically all other PRs? @qbittorrent/frequent-contributors?

glassez commented:
>When I update this PR now, do you think I should move its files into gui/widgets or not? 

Not here, exactly.

Chocobo1 commented:
>When I update this PR now, do you think I should move its files into gui/widgets or not? Should I submit another PR instead to rearrange all files (gui/models, gui/views, gui/widgets) and break basically all other PRs?

Or you can wait till most existing PRs are merged, we are merging fast recently...

evsh commented:
OK, thank you for your input! Then for this one the files will be kept at the same place.

evsh commented:
PR updated:
- Fixed filter string for IP filters in options dialog (@Chocobo1)
- Widget classes migrated to Qt5: removed `#ifdef`s and new signal/slot syntax is used. Private slot is replaced with lambda.
- Various formatting issues were corrected.

glassez commented on src/gui/fspathedit.h:
```diff
@@ -0,0 +1,151 @@
+
+    virtual void clear() = 0;
+
+	signals:
```
Broken indentation. And next line too.

glassez commented on src/gui/fspathedit_p.cpp:
```diff
@@ -0,0 +1,133 @@
+    m_completer->setCompletionMode(QCompleter::PopupCompletion);
+    setCompleter(m_completer);
+
+    connect(m_completerModel, SIGNAL(directoryLoaded(QString)), this, SLOT(showCompletionPopup()));
```
Please, use Qt5 style here too.

Chocobo1 commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,361 @@
+    m_browseAction->setIconText(browseButtonBriefText.tr());
+    m_browseAction->setText(browseButtonFullText.tr());
+    m_browseAction->setToolTip(browseButtonFullText.tr().remove(QLatin1Char('&')));
+    m_browseAction->setShortcut({Qt::CTRL + Qt::Key_B});
```
curly braces required?

Chocobo1 commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,361 @@
+
+void FileSystemPathEdit::FileSystemPathEditPrivate::browseActionTriggered()
+{
+    QString selectedPath;
```
I would move this to just above switch expression.

Chocobo1 commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,361 @@
+        pixmap = QStyle::SP_DirOpenIcon;
+        showDirsOnly = true;
+        break;
+    }
```
I'm not sure if a default case would make sense... but it doesn't hurt to add one?

evsh replied:
Oh, yes! `default` with `throw` was mistakenly forgotten here.

Chocobo1 commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,361 @@
+    Q_D(FileSystemPathEdit);
+    d->m_fileNameFilter = val;
+
+#if 0
```
`#if 0` for comments? `/**/` does not suffice?

evsh replied:
Yes, `#if 0` is perfect for commenting out a block of code which itself contains comments inside.

Chocobo1 commented on src/gui/fspathedit_p.h:
```diff
@@ -0,0 +1,101 @@
+#include <QFileSystemModel>
+#include <QKeyEvent>
+#include <QLineEdit>
+#include <QMenu>
```
forward declare as much as possible?

evsh replied:
I don't think it is important here, because this is a private header, included from only single other place.

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -357,6 +353,25 @@ OptionsDialog::OptionsDialog(QWidget *parent)
+
+    m_ui->textFilterPath->setDialogCaption(tr("Choose an IP filter file"));
+    m_ui->textFilterPath->setFileNameFilter(tr("All supported filters")
+        + QString(" (*.dat *.p2p *.p2b);;.dat (*.dat);;.p2p (*.p2p);;.p2b (*.p2b)"));
```
`QString` -> `QLatin1String` ?

evsh commented:
Comments from reviewers are addressed.

glassez commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -34,6 +34,7 @@
 #include <QUrl>
 #include <QMenu>
 #include <QFileDialog>
+#include <QPushButton>
```
Is this include needed?

evsh replied:
Yes, it is needed. Otherwise:
```
../../src/gui/addnewtorrentdialog.cpp:148:45: error: invalid use of incomplete type ‘class QPushButton’
  ui->buttonBox->button(QDialogButtonBox::Ok)->setFocus();
                                             ^~
```

glassez commented on src/gui/fspathedit.cpp:
```diff
@@ -0,0 +1,361 @@
+    QString m_dialogCaption;
+};
+
+FileSystemPathEdit::FileSystemPathEditPrivate::FileSystemPathEditPrivate(FileSystemPathEdit *q,
```
IMO, it would be better if you insert line break before first param (instead of second one).

evsh replied:
Taken.

glassez commented on src/gui/optionsdlg.cpp:
```diff
@@ -357,6 +353,25 @@ OptionsDialog::OptionsDialog(QWidget *parent)
     m_ui->advPageLayout->addWidget(advancedSettings);
     connect(advancedSettings, SIGNAL(settingsChanged()), this, SLOT(enableApplyButton()));
 
+    m_ui->textFileLogPath->setDialogCaption(tr("Choose a save directory"));
```
Can't you set these properties in designer?

evsh replied:
Nope, I do not export the widget to Designer completely. Probably, with only a single custom widget in the app it does not make sense.

evsh commented:
PR updated: taken two last suggestions from the reviewers. @sledgehammer999, do you want to say anything?

evsh commented:
@sledgehammer999, should I count a week? Are you interested in this?

evsh commented:
OK, a week has passed, merging.

Chocobo1 commented:
Might be related to this PR, I got the following when opening options dialog:
```
QObject::connect: No such signal FileSystemPathEdit::textChanged(QString)
QObject::connect:  (sender name:   'textFilterPath')
QObject::connect:  (receiver name: 'OptionsDialog')
```

glassez commented:
Damn, we should prohibit using old connect syntax in new code to prevent this bugs in future.

evsh commented:
I'm sorry for that. See #6795, please.

Chocobo1 commented on src/gui/optionsdlg.ui:
```diff
@@ -6,8 +6,8 @@
-    <width>779</width>
-    <height>591</height>
+    <width>1116</width>
+    <height>838</height>
```
This is causing the default dialog size to be large than ever, is it intentional?
Do you mind we switch back?

hrastnik commented:
Did you consider the cases where the users manually writes invalid file paths?

evsh commented:
@Chocobo1: sure
@hrastnik: no, I did not.
